File: day13/aws-vpc-python/private_subnet.py
import pulumi
import pulumi_aws as aws


class PrivateSubnet(pulumi.ComponentResource):
    def __init__(self, name, opts: pulumi.ResourceOptions = None, **kwargs):
        super().__init__('aws-vpc-python:private_subnet:PrivateSubnet', name, kwargs, opts)
        az = kwargs['az']
        cidr = kwargs['cidr_block']
        vpc_id = kwargs['vpc_id']

        self.name = name

        self.subnet = aws.ec2.Subnet(self.name,
                                     vpc_id=vpc_id,
                                     cidr_block=cidr,
                                     availability_zone=az,
                                     opts=pulumi.ResourceOptions(parent=self))

        self.rt = aws.ec2.RouteTable(f'{self.name}-rt',
                                     vpc_id=vpc_id,
                                     opts=pulumi.ResourceOptions(parent=self))
        self.default_route = None

        aws.ec2.RouteTableAssociation(f'{self.name}-rt-association',
                                      subnet_id=self.subnet.id,
                                      route_table_id=self.rt.id,
                                      opts=pulumi.ResourceOptions(parent=self))

    def add_nat_gateway(self, nat_gateway_id):
        if self.default_route is not None:
            raise Exception('Default route already exists')

        self.default_route = aws.ec2.Route(f'{self.name}-default-route',
                                           route_table_id=self.rt.id,
                                           destination_cidr_block='0.0.0.0/0',
                                           nat_gateway_id=nat_gateway_id,
                                           opts=pulumi.ResourceOptions(parent=self))

File: day06/day06-aws-vpc-ts/privateSubnet.ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

interface CreatePrivateSubnetArgs {
  vpcId: pulumi.Input<string>;
  cidr: string;
  az: string;
  defaultTags: Record<string, string>;
}

export class PrivateSubnet {
  public name: string;
  public subnet: aws.ec2.Subnet;

  private routeTable: aws.ec2.RouteTable;

  private defaultRoute?: aws.ec2.Route;

  constructor(name: string, args: CreatePrivateSubnetArgs) {
    this.name = name;

    this.subnet = new aws.ec2.Subnet(this.name, {
      vpcId: args.vpcId,
      cidrBlock: args.cidr,
      availabilityZone: args.az,
      tags: {
        "Name": this.name,
        ...args.defaultTags
      }
    });

    this.routeTable = new aws.ec2.RouteTable(`${this.name}-rt`, {
      vpcId: args.vpcId,
      tags: {
        'Name': `${this.name}-rt`,
        ...args.defaultTags
      }
    });

    new aws.ec2.RouteTableAssociation(`${this.name}-rt-association`, {
      routeTableId: this.routeTable.id,
      subnetId: this.subnet.id,
    });
  }

  addNatGateway(natGatewayId: pulumi.Input<string>) {
    if (this.defaultRoute) {
      throw new Error('Cannot add a NAT gateway to a subnet that already has a default route');
    }
    this.defaultRoute = new aws.ec2.Route(`${this.name}-nat-gateway-route`, {
      routeTableId: this.routeTable.id,
      destinationCidrBlock: '0.0.0.0/0',
      natGatewayId: natGatewayId,
    });
  }
}
